src/boot/efi/devicetree.c

```diff
@@ -0,0 +1,185 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <efi.h>
+
+#include "devicetree.h"
+#include "util.h"
+
+#define FDT_V1_SIZE (7*4)
+
+#ifndef DEVICE_TREE_GUID
+#define DEVICE_TREE_GUID \
+        { 0xb1b621d5, 0xf19c, 0x41a5, {0x83, 0x0b, 0xd9, 0x15, 0x2c, 0x69, 0xaa, 0xe0} }
+#endif
+static const EFI_GUID fdt_guid = DEVICE_TREE_GUID;
+
+#ifndef EFI_DT_FIXUP_PROTOCOL_GUID
+#define EFI_DT_FIXUP_PROTOCOL_GUID \
+        { 0xe617d64c, 0xfe08, 0x46da, {0xf4, 0xdc, 0xbb, 0xd5, 0x87, 0x0c, 0x73, 0x00} }
+#endif
+static const EFI_GUID dt_fixup_guid = EFI_DT_FIXUP_PROTOCOL_GUID;
+
+#define EFI_DT_FIXUP_PROTOCOL_REVISION 0x00010000
+
+/* Add nodes and update properties */
+#define EFI_DT_APPLY_FIXUPS    0x00000001
+/*
+ * Reserve memory according to the /reserved-memory node
+ * and the memory reservation block
+ */
+#define EFI_DT_RESERVE_MEMORY  0x00000002
+
+typedef struct _EFI_DT_FIXUP_PROTOCOL EFI_DT_FIXUP_PROTOCOL;
+
+typedef EFI_STATUS (EFIAPI *EFI_DT_FIXUP) (
+        IN EFI_DT_FIXUP_PROTOCOL *This,
+        IN VOID                  *Fdt,
+        IN OUT UINTN             *BufferSize,
+        IN UINT32                Flags);
+
+struct _EFI_DT_FIXUP_PROTOCOL {
+        UINT64         Revision;
+        EFI_DT_FIXUP   Fixup;
+};
+
+static EFI_STATUS devicetree_get(struct devicetree_state *state) {
```

Comment:
We have assert() now, please use it.

Reply:
What exactly is it that needs to be asserted here?

Reply:
That pointers passed to functions aren't NULL.

Reply:
The cleanup function too, or are we expecting the compiler to behave?

Reply:
So far the cleanup functions don't assert there, so no.